usr/sbin/pkcsslotd/slotmgr.c

```diff
@@ -165,15 +202,41 @@ int drop_capabilities(void)
  */
 void drop_privileges(struct passwd *pwd)
 {
-    char program[PATH_MAX + 1] = { 0 };
-    char* args[] = { program, NULL };
+    char program[PATH_MAX + 1] = { 0, };
+
+#if defined(_AIX)
+    struct psinfo ps;
+    char procpath[PATH_MAX + 1] = { NULL, };
+
+    snprintf(procpath, PATH_MAX - 13, "/proc/%lld/psinfo", getpid());
+    int psfd = open(procpath, O_RDONLY);
+    if (psfd == -1) {
+        fprintf(stderr,
+            "Failed to open procfs to read cmdname: %s\n", strerror(errno));
+        exit(-1);
+    }
+
+    if (read(psfd, &ps, sizeof(ps)) == -1) {
+        fprintf(stderr, "Failed to populate psinfo: %s\n", strerror(errno));
+        exit(-1);
+    }
+    close(psfd);
+    /*
+     * Copies only argv[0], as maintained by the kernel. This is because
+     * parameters are guaranteed to be separated by NULL, which is where strcpy
+     * stops copying.
+     */
+    strncpy(program, ps.pr_psargs, PATH_MAX);
 
+#else
     if (readlink("/proc/self/exe", program, PATH_MAX) == -1) {
         fprintf(stderr, "Failed to get executable file name: %s\n",
                 strerror(errno));
         exit(1);
     }
+#endif
 
+    char* args[] = { program, NULL };
```

Comment:
please move this back to above the code where it was, right after the variable declaration of program.